File: shared/protocol.ts
/**
 * Robot Control Protocol - Shared Type Definitions
 *
 * This file defines the message schemas for communication between:
 * - Client (browser) - source of truth for desired state (inputs)
 * - Server (API) - relay, fan-out, persistence
 * - Robot - source of truth for actual state (sensors)
 *
 * All messages are JSON-encoded and wrapped in a MessageEnvelope.
 */

// ============================================================================
// MESSAGE ENVELOPE
// ============================================================================

export interface MessageEnvelope {
  type: MessageType;
  robot_id: string;
  timestamp: number;  // Unix timestamp in milliseconds
  seq?: number;       // Optional sequence number for ordering
  payload: unknown;   // Type depends on message type
}

export type MessageType =
  // Client → Robot (via server)
  | 'controller_state'
  | 'command'
  // Robot → Client (via server)
  | 'robot_state'
  | 'telemetry'
  | 'map_update'
  | 'error'
  // Server management
  | 'subscribe'
  | 'unsubscribe'
  | 'heartbeat';

// ============================================================================
// CONTROLLER STATE (Client → Robot)
// Normalized gamepad-style input that works across all input devices
// ============================================================================

export interface ControllerState {
  // Analog axes (-1.0 to 1.0)
  axes: {
    left_x: number;       // Left stick horizontal (strafe/steering)
    left_y: number;       // Left stick vertical (forward/back)
    right_x: number;      // Right stick horizontal (camera pan/turn)
    right_y: number;      // Right stick vertical (camera tilt)
    left_trigger: number; // 0.0 to 1.0 (brake/reverse modifier)
    right_trigger: number; // 0.0 to 1.0 (throttle/boost)
  };

  // Digital buttons (true = pressed)
  buttons: {
    a: boolean;           // Primary action
    b: boolean;           // Secondary action / cancel
    x: boolean;           // Tertiary action
    y: boolean;           // Quaternary action
    left_bumper: boolean;
    right_bumper: boolean;
    back: boolean;        // Select/back
    start: boolean;       // Start/menu
    left_stick: boolean;  // L3
    right_stick: boolean; // R3
    dpad_up: boolean;
    dpad_down: boolean;
    dpad_left: boolean;
    dpad_right: boolean;
  };

  // Input source metadata
  source: InputSource;
}

export type InputSource =
  | 'keyboard'
  | 'gamepad'
  | 'touch'      // Virtual joystick
  | 'tilt'       // Phone accelerometer
  | 'mouse';

// Default/zero controller state
export const ZERO_CONTROLLER_STATE: ControllerState = {
  axes: {
    left_x: 0, left_y: 0,
    right_x: 0, right_y: 0,
    left_trigger: 0, right_trigger: 0,
  },
  buttons: {
    a: false, b: false, x: false, y: false,
    left_bumper: false, right_bumper: false,
    back: false, start: false,
    left_stick: false, right_stick: false,
    dpad_up: false, dpad_down: false,
    dpad_left: false, dpad_right: false,
  },
  source: 'keyboard',
};

// ============================================================================
// ROBOT STATE (Robot → Client)
// Actual state from sensors and hardware
// ============================================================================

export interface RobotState {
  // Identity
  robot_id: string;
  robot_type: string;  // e.g., 'rosmaster_a1', 'turtlebot3', 'drone'

  // Connection
  connected: boolean;
  last_seen: number;  // Unix timestamp ms

  // Pose (from odometry/SLAM)
  pose: Pose3D;

  // Velocity (current motion)
  velocity: Velocity;

  // Hardware state
  servos: ServoState;
  battery: BatteryState;

  // Status
  mode: RobotMode;
  errors: string[];
}

export interface Pose3D {
  // Position (meters, relative to origin/map frame)
  x: number;
  y: number;
  z: number;

  // Orientation (radians)
  roll: number;   // Rotation around X
  pitch: number;  // Rotation around Y
  yaw: number;    // Rotation around Z (heading)
}

export interface Velocity {
  // Linear velocity (m/s)
  linear_x: number;  // Forward/back
  linear_y: number;  // Left/right (for omnidirectional)
  linear_z: number;  // Up/down (for drones)

  // Angular velocity (rad/s)
  angular_x: number;  // Roll rate
  angular_y: number;  // Pitch rate
  angular_z: number;  // Yaw rate (turn rate)
}

export interface ServoState {
  // Camera servos (degrees, 0-180)
  camera_pan: number;
  camera_tilt: number;

  // Steering (for Ackermann, degrees or normalized)
  steering_angle?: number;

  // Additional servos (robot-specific)
  [key: string]: number | undefined;
}

export interface BatteryState {
  voltage: number;        // Volts
  percentage?: number;    // 0-100 if known
  charging?: boolean;
  cells?: number;         // Number of cells (e.g., 3S = 3)
  low_warning?: boolean;  // True if below threshold
}

export type RobotMode =
  | 'idle'
  | 'manual'        // Human control via controller
  | 'autonomous'    // Self-driving / following waypoints
  | 'mapping'       // SLAM mapping mode
  | 'docking'       // Returning to dock
  | 'emergency_stop';

// ============================================================================
// TELEMETRY (Robot → Client)
// High-frequency sensor updates (sent separately from full state)
// ============================================================================

export interface Telemetry {
  timestamp: number;

  // IMU data
  imu?: IMUData;

  // Wheel encoders
  encoders?: EncoderData;

  // Distance sensors
  distance_sensors?: DistanceSensorData;
}

export interface IMUData {
  // Acceleration (m/s²)
  accel_x: number;
  accel_y: number;
  accel_z: number;

  // Gyroscope (rad/s)
  gyro_x: number;
  gyro_y: number;
  gyro_z: number;

  // Magnetometer (if available)
  mag_x?: number;
  mag_y?: number;
  mag_z?: number;
}

export interface EncoderData {
  // Wheel velocities (rad/s or m/s depending on robot)
  left: number;
  right: number;
  // For omnidirectional
  front_left?: number;
  front_right?: number;
  rear_left?: number;
  rear_right?: number;
}

export interface DistanceSensorData {
  // Distance readings (meters), undefined if no obstacle detected
  front?: number;
  rear?: number;
  left?: number;
  right?: number;
  // Array for LIDAR scan (if available)
  scan?: number[];
  scan_angle_min?: number;  // radians
  scan_angle_max?: number;
  scan_angle_increment?: number;
}

// ============================================================================
// MAP DATA (Robot → Client)
// SLAM occupancy grid, sent on change or periodically
// ============================================================================

export interface MapUpdate {
  // Map metadata
  resolution: number;  // Meters per cell
  width: number;       // Cells
  height: number;      // Cells

  // Origin (position of cell 0,0 in world frame)
  origin: {
    x: number;
    y: number;
    yaw: number;
  };

  // Occupancy data: -1 = unknown, 0 = free, 100 = occupied
  // Encoded as base64 for efficiency, or as array for simplicity
  data: number[] | string;
  encoding?: 'raw' | 'base64' | 'rle';  // Run-length encoding for compression

  // Incremental update (if only part of map changed)
  incremental?: boolean;
  region?: { x: number; y: number; width: number; height: number };
}

// ============================================================================
// COMMANDS (Client → Robot)
// Discrete commands (not continuous control)
// ============================================================================

export interface Command {
  id: string;          // Unique command ID for acknowledgment
  action: CommandAction;
  params?: Record<string, unknown>;
}

export type CommandAction =
  | 'set_mode'         // params: { mode: RobotMode }
  | 'calibrate'        // params: { sensor: string }
  | 'home'             // Return to origin
  | 'dock'             // Return to charging dock
  | 'emergency_stop'
  | 'clear_errors'
  | 'set_waypoint'     // params: { x, y, yaw }
  | 'cancel_waypoint'
  | 'start_mapping'
  | 'stop_mapping'
  | 'save_map'         // params: { name: string }
  | 'load_map'         // params: { name: string }
  | 'beep'             // params: { duration_ms: number }
  | 'set_led';         // params: { r, g, b } or { pattern: string }

export interface CommandAck {
  command_id: string;
  status: 'received' | 'executing' | 'completed' | 'failed';
  error?: string;
}

// ============================================================================
// KEYBOARD MAPPING
// Default keyboard to controller state mapping
// ============================================================================

export interface KeyMapping {
  key: string;
  axis?: keyof ControllerState['axes'];
  axisValue?: number;  // Value when pressed (-1, 1, etc.)
  button?: keyof ControllerState['buttons'];
}

export const DEFAULT_KEY_MAPPINGS: KeyMapping[] = [
  // WASD → Left stick
  { key: 'w', axis: 'left_y', axisValue: 1.0 },
  { key: 's', axis: 'left_y', axisValue: -1.0 },
  { key: 'a', axis: 'left_x', axisValue: -1.0 },
  { key: 'd', axis: 'left_x', axisValue: 1.0 },

  // Arrow keys → Right stick
  { key: 'ArrowUp', axis: 'right_y', axisValue: 1.0 },
  { key: 'ArrowDown', axis: 'right_y', axisValue: -1.0 },
  { key: 'ArrowLeft', axis: 'right_x', axisValue: -1.0 },
  { key: 'ArrowRight', axis: 'right_x', axisValue: 1.0 },

  // Triggers
  { key: 'Shift', axis: 'left_trigger', axisValue: 1.0 },
  { key: ' ', axis: 'right_trigger', axisValue: 1.0 },  // Space

  // Buttons
  { key: 'e', button: 'a' },
  { key: 'q', button: 'b' },
  { key: 'r', button: 'x' },
  { key: 'f', button: 'y' },
  { key: 'Tab', button: 'back' },
  { key: 'Escape', button: 'start' },
  { key: '1', button: 'dpad_left' },
  { key: '2', button: 'dpad_up' },
  { key: '3', button: 'dpad_down' },
  { key: '4', button: 'dpad_right' },
];

// ============================================================================
// GAMEPAD MAPPING
// Standard gamepad axis/button indices to our named fields
// Based on "Standard Gamepad" mapping (Xbox-style layout)
// https://w3c.github.io/gamepad/#remapping
// ============================================================================

export const STANDARD_GAMEPAD_MAPPING = {
  axes: {
    0: 'left_x',      // Left stick X
    1: 'left_y',      // Left stick Y (often inverted)
    2: 'right_x',     // Right stick X
    3: 'right_y',     // Right stick Y (often inverted)
  },
  buttons: {
    0: 'a',           // A (bottom)
    1: 'b',           // B (right)
    2: 'x',           // X (left)
    3: 'y',           // Y (top)
    4: 'left_bumper',
    5: 'right_bumper',
    6: 'left_trigger',  // Analog, use value
    7: 'right_trigger', // Analog, use value
    8: 'back',
    9: 'start',
    10: 'left_stick',
    11: 'right_stick',
    12: 'dpad_up',
    13: 'dpad_down',
    14: 'dpad_left',
    15: 'dpad_right',
  },
  // Axes that need Y inversion (push up = negative by default)
  invertY: [1, 3],
} as const;

File: shared/protocol.py
"""
Robot Control Protocol - Shared Type Definitions (Python)

This file defines the message schemas for communication between:
- Client (browser) - source of truth for desired state (inputs)
- Server (API) - relay, fan-out, persistence
- Robot - source of truth for actual state (sensors)

All messages are JSON-encoded and wrapped in a MessageEnvelope.
"""

from dataclasses import dataclass, field
from enum import Enum
from typing import Dict, List, Literal, Optional, Union
import time


# ============================================================================
# ENUMS
# ============================================================================

class MessageType(str, Enum):
    # Client → Robot (via server)
    CONTROLLER_STATE = "controller_state"
    COMMAND = "command"
    # Robot → Client (via server)
    ROBOT_STATE = "robot_state"
    TELEMETRY = "telemetry"
    MAP_UPDATE = "map_update"
    ERROR = "error"
    # Server management
    SUBSCRIBE = "subscribe"
    UNSUBSCRIBE = "unsubscribe"
    HEARTBEAT = "heartbeat"
    # Virtual World (Tauri ↔ API)
    GET_WORLD_STATE = "get_world_state"
    WORLD_STATE = "world_state"
    ADD_ELEMENT = "add_element"
    UPDATE_ELEMENT = "update_element"
    REMOVE_ELEMENT = "remove_element"
    ELEMENT_ADDED = "element_added"
    ELEMENT_UPDATED = "element_updated"
    ELEMENT_REMOVED = "element_removed"
    PLAYER_STATE = "player_state"
    VIRTUAL_PLAYER_CREATED = "virtual_player_created"
    VIRTUAL_PLAYER_DELETED = "virtual_player_deleted"


class InputSource(str, Enum):
    KEYBOARD = "keyboard"
    GAMEPAD = "gamepad"
    TOUCH = "touch"
    TILT = "tilt"
    MOUSE = "mouse"


class RobotMode(str, Enum):
    IDLE = "idle"
    MANUAL = "manual"
    AUTONOMOUS = "autonomous"
    MAPPING = "mapping"
    DOCKING = "docking"
    EMERGENCY_STOP = "emergency_stop"


class CommandAction(str, Enum):
    SET_MODE = "set_mode"
    CALIBRATE = "calibrate"
    HOME = "home"
    DOCK = "dock"
    EMERGENCY_STOP = "emergency_stop"
    CLEAR_ERRORS = "clear_errors"
    SET_WAYPOINT = "set_waypoint"
    CANCEL_WAYPOINT = "cancel_waypoint"
    START_MAPPING = "start_mapping"
    STOP_MAPPING = "stop_mapping"
    SAVE_MAP = "save_map"
    LOAD_MAP = "load_map"
    BEEP = "beep"
    SET_LED = "set_led"


# ============================================================================
# CONTROLLER STATE (Client → Robot)
# ============================================================================

@dataclass
class ControllerAxes:
    """Analog axes, all values -1.0 to 1.0 (triggers 0.0 to 1.0)"""
    left_x: float = 0.0
    left_y: float = 0.0
    right_x: float = 0.0
    right_y: float = 0.0
    left_trigger: float = 0.0
    right_trigger: float = 0.0

    def to_dict(self) -> Dict[str, float]:
        return {
            "left_x": self.left_x,
            "left_y": self.left_y,
            "right_x": self.right_x,
            "right_y": self.right_y,
            "left_trigger": self.left_trigger,
            "right_trigger": self.right_trigger,
        }

    @classmethod
    def from_dict(cls, d: Dict) -> "ControllerAxes":
        return cls(
            left_x=d.get("left_x", 0.0),
            left_y=d.get("left_y", 0.0),
            right_x=d.get("right_x", 0.0),
            right_y=d.get("right_y", 0.0),
            left_trigger=d.get("left_trigger", 0.0),
            right_trigger=d.get("right_trigger", 0.0),
        )


@dataclass
class ControllerButtons:
    """Digital buttons, all boolean"""
    a: bool = False
    b: bool = False
    x: bool = False
    y: bool = False
    left_bumper: bool = False
    right_bumper: bool = False
    back: bool = False
    start: bool = False
    left_stick: bool = False
    right_stick: bool = False
    dpad_up: bool = False
    dpad_down: bool = False
    dpad_left: bool = False
    dpad_right: bool = False

    def to_dict(self) -> Dict[str, bool]:
        return {
            "a": self.a,
            "b": self.b,
            "x": self.x,
            "y": self.y,
            "left_bumper": self.left_bumper,
            "right_bumper": self.right_bumper,
            "back": self.back,
            "start": self.start,
            "left_stick": self.left_stick,
            "right_stick": self.right_stick,
            "dpad_up": self.dpad_up,
            "dpad_down": self.dpad_down,
            "dpad_left": self.dpad_left,
            "dpad_right": self.dpad_right,
        }

    @classmethod
    def from_dict(cls, d: Dict) -> "ControllerButtons":
        return cls(**{k: d.get(k, False) for k in cls.__dataclass_fields__})


@dataclass
class ControllerState:
    """Normalized gamepad-style input from any input device"""
    axes: ControllerAxes = field(default_factory=ControllerAxes)
    buttons: ControllerButtons = field(default_factory=ControllerButtons)
    source: InputSource = InputSource.KEYBOARD

    def to_dict(self) -> Dict:
        return {
            "axes": self.axes.to_dict(),
            "buttons": self.buttons.to_dict(),
            "source": self.source.value,
        }

    @classmethod
    def from_dict(cls, d: Dict) -> "ControllerState":
        return cls(
            axes=ControllerAxes.from_dict(d.get("axes", {})),
            buttons=ControllerButtons.from_dict(d.get("buttons", {})),
            source=InputSource(d.get("source", "keyboard")),
        )


# ============================================================================
# ROBOT STATE (Robot → Client)
# ============================================================================

@dataclass
class Pose3D:
    """Position and orientation in 3D space"""
    x: float = 0.0
    y: float = 0.0
    z: float = 0.0
    roll: float = 0.0
    pitch: float = 0.0
    yaw: float = 0.0

    def to_dict(self) -> Dict:
        return {
            "x": self.x, "y": self.y, "z": self.z,
            "roll": self.roll, "pitch": self.pitch, "yaw": self.yaw,
        }

    @classmethod
    def from_dict(cls, d: Dict) -> "Pose3D":
        return cls(**{k: d.get(k, 0.0) for k in cls.__dataclass_fields__})


@dataclass
class Velocity:
    """Linear and angular velocities"""
    linear_x: float = 0.0
    linear_y: float = 0.0
    linear_z: float = 0.0
    angular_x: float = 0.0
    angular_y: float = 0.0
    angular_z: float = 0.0

    def to_dict(self) -> Dict:
        return {
            "linear_x": self.linear_x,
            "linear_y": self.linear_y,
            "linear_z": self.linear_z,
            "angular_x": self.angular_x,
            "angular_y": self.angular_y,
            "angular_z": self.angular_z,
        }

    @classmethod
    def from_dict(cls, d: Dict) -> "Velocity":
        return cls(**{k: d.get(k, 0.0) for k in cls.__dataclass_fields__})


@dataclass
class ServoState:
    """Servo positions (typically camera pan/tilt)"""
    camera_pan: float = 90.0   # degrees, 0-180
    camera_tilt: float = 90.0  # degrees, 0-180
    steering_angle: Optional[float] = None  # For Ackermann robots
    extra: Dict[str, float] = field(default_factory=dict)

    def to_dict(self) -> Dict:
        d = {
            "camera_pan": self.camera_pan,
            "camera_tilt": self.camera_tilt,
        }
        if self.steering_angle is not None:
            d["steering_angle"] = self.steering_angle
        d.update(self.extra)
        return d

    @classmethod
    def from_dict(cls, d: Dict) -> "ServoState":
        return cls(
            camera_pan=d.get("camera_pan", 90.0),
            camera_tilt=d.get("camera_tilt", 90.0),
            steering_angle=d.get("steering_angle"),
            extra={k: v for k, v in d.items()
                   if k not in ("camera_pan", "camera_tilt", "steering_angle")},
        )


@dataclass
class BatteryState:
    """Battery status"""
    voltage: float = 0.0
    percentage: Optional[float] = None
    charging: Optional[bool] = None
    cells: Optional[int] = None
    low_warning: bool = False

    def to_dict(self) -> Dict:
        d = {"voltage": self.voltage, "low_warning": self.low_warning}
        if self.percentage is not None:
            d["percentage"] = self.percentage
        if self.charging is not None:
            d["charging"] = self.charging
        if self.cells is not None:
            d["cells"] = self.cells
        return d

    @classmethod
    def from_dict(cls, d: Dict) -> "BatteryState":
        return cls(
            voltage=d.get("voltage", 0.0),
            percentage=d.get("percentage"),
            charging=d.get("charging"),
            cells=d.get("cells"),
            low_warning=d.get("low_warning", False),
        )


@dataclass
class RobotState:
    """Full robot state snapshot"""
    robot_id: str = ""
    robot_type: str = ""
    connected: bool = False
    last_seen: float = 0.0
    pose: Pose3D = field(default_factory=Pose3D)
    velocity: Velocity = field(default_factory=Velocity)
    servos: ServoState = field(default_factory=ServoState)
    battery: BatteryState = field(default_factory=BatteryState)
    mode: RobotMode = RobotMode.IDLE
    errors: List[str] = field(default_factory=list)

    def to_dict(self) -> Dict:
        return {
            "robot_id": self.robot_id,
            "robot_type": self.robot_type,
            "connected": self.connected,
            "last_seen": self.last_seen,
            "pose": self.pose.to_dict(),
            "velocity": self.velocity.to_dict(),
            "servos": self.servos.to_dict(),
            "battery": self.battery.to_dict(),
            "mode": self.mode.value,
            "errors": self.errors,
        }

    @classmethod
    def from_dict(cls, d: Dict) -> "RobotState":
        return cls(
            robot_id=d.get("robot_id", ""),
            robot_type=d.get("robot_type", ""),
            connected=d.get("connected", False),
            last_seen=d.get("last_seen", 0.0),
            pose=Pose3D.from_dict(d.get("pose", {})),
            velocity=Velocity.from_dict(d.get("velocity", {})),
            servos=ServoState.from_dict(d.get("servos", {})),
            battery=BatteryState.from_dict(d.get("battery", {})),
            mode=RobotMode(d.get("mode", "idle")),
            errors=d.get("errors", []),
        )


# ============================================================================
# TELEMETRY (Robot → Client)
# ============================================================================

@dataclass
class IMUData:
    """Inertial measurement unit data"""
    accel_x: float = 0.0
    accel_y: float = 0.0
    accel_z: float = 0.0
    gyro_x: float = 0.0
    gyro_y: float = 0.0
    gyro_z: float = 0.0
    mag_x: Optional[float] = None
    mag_y: Optional[float] = None
    mag_z: Optional[float] = None

    def to_dict(self) -> Dict:
        d = {
            "accel_x": self.accel_x, "accel_y": self.accel_y, "accel_z": self.accel_z,
            "gyro_x": self.gyro_x, "gyro_y": self.gyro_y, "gyro_z": self.gyro_z,
        }
        if self.mag_x is not None:
            d.update({"mag_x": self.mag_x, "mag_y": self.mag_y, "mag_z": self.mag_z})
        return d

    @classmethod
    def from_dict(cls, d: Dict) -> "IMUData":
        return cls(
            accel_x=d.get("accel_x", 0.0),
            accel_y=d.get("accel_y", 0.0),
            accel_z=d.get("accel_z", 0.0),
            gyro_x=d.get("gyro_x", 0.0),
            gyro_y=d.get("gyro_y", 0.0),
            gyro_z=d.get("gyro_z", 0.0),
            mag_x=d.get("mag_x"),
            mag_y=d.get("mag_y"),
            mag_z=d.get("mag_z"),
        )


@dataclass
class Telemetry:
    """High-frequency sensor updates"""
    timestamp: float = field(default_factory=lambda: time.time() * 1000)
    imu: Optional[IMUData] = None
    encoders: Optional[Dict[str, float]] = None
    distance_sensors: Optional[Dict[str, float]] = None

    def to_dict(self) -> Dict:
        d = {"timestamp": self.timestamp}
        if self.imu:
            d["imu"] = self.imu.to_dict()
        if self.encoders:
            d["encoders"] = self.encoders
        if self.distance_sensors:
            d["distance_sensors"] = self.distance_sensors
        return d

    @classmethod
    def from_dict(cls, d: Dict) -> "Telemetry":
        return cls(
            timestamp=d.get("timestamp", time.time() * 1000),
            imu=IMUData.from_dict(d["imu"]) if "imu" in d else None,
            encoders=d.get("encoders"),
            distance_sensors=d.get("distance_sensors"),
        )


# ============================================================================
# MAP DATA (Robot → Client)
# ============================================================================

@dataclass
class MapUpdate:
    """SLAM occupancy grid update"""
    resolution: float = 0.05  # meters per cell
    width: int = 0
    height: int = 0
    origin_x: float = 0.0
    origin_y: float = 0.0
    origin_yaw: float = 0.0
    data: Union[List[int], str] = field(default_factory=list)
    encoding: Literal["raw", "base64", "rle"] = "raw"
    incremental: bool = False
    region: Optional[Dict[str, int]] = None  # {x, y, width, height}

    def to_dict(self) -> Dict:
        d = {
            "resolution": self.resolution,
            "width": self.width,
            "height": self.height,
            "origin": {
                "x": self.origin_x,
                "y": self.origin_y,
                "yaw": self.origin_yaw,
            },
            "data": self.data,
            "encoding": self.encoding,
            "incremental": self.incremental,
        }
        if self.region:
            d["region"] = self.region
        return d


# ============================================================================
# COMMANDS (Client → Robot)
# ============================================================================

@dataclass
class Command:
    """Discrete command (not continuous control)"""
    id: str
    action: CommandAction
    params: Dict = field(default_factory=dict)

    def to_dict(self) -> Dict:
        return {
            "id": self.id,
            "action": self.action.value,
            "params": self.params,
        }

    @classmethod
    def from_dict(cls, d: Dict) -> "Command":
        return cls(
            id=d.get("id", ""),
            action=CommandAction(d.get("action", "beep")),
            params=d.get("params", {}),
        )


@dataclass
class CommandAck:
    """Command acknowledgment"""
    command_id: str
    status: Literal["received", "executing", "completed", "failed"]
    error: Optional[str] = None

    def to_dict(self) -> Dict:
        d = {"command_id": self.command_id, "status": self.status}
        if self.error:
            d["error"] = self.error
        return d


# ============================================================================
# MESSAGE ENVELOPE
# ============================================================================

@dataclass
class MessageEnvelope:
    """Wrapper for all protocol messages"""
    type: MessageType
    robot_id: str
    timestamp: float = field(default_factory=lambda: time.time() * 1000)
    seq: Optional[int] = None
    payload: Dict = field(default_factory=dict)

    def to_dict(self) -> Dict:
        d = {
            "type": self.type.value,
            "robot_id": self.robot_id,
            "timestamp": self.timestamp,
            "payload": self.payload,
        }
        if self.seq is not None:
            d["seq"] = self.seq
        return d

    @classmethod
    def from_dict(cls, d: Dict) -> "MessageEnvelope":
        return cls(
            type=MessageType(d.get("type", "heartbeat")),
            robot_id=d.get("robot_id", ""),
            timestamp=d.get("timestamp", time.time() * 1000),
            seq=d.get("seq"),
            payload=d.get("payload", {}),
        )


# ============================================================================
# ROBOT CONTROL PROFILE
# Maps controller state to robot-specific actions
# ============================================================================

@dataclass
class RobotControlProfile:
    """
    Defines how a specific robot type interprets controller input.
    Each robot type can define its own mapping.
    """
    robot_type: str

    # Axis mappings: controller axis -> robot action
    # e.g., {"left_y": "throttle", "left_x": "steering"}
    axis_mappings: Dict[str, str] = field(default_factory=dict)

    # Axis scales and limits
    axis_scales: Dict[str, float] = field(default_factory=dict)  # multipliers
    axis_limits: Dict[str, tuple] = field(default_factory=dict)  # (min, max)

    # Button mappings: controller button -> command action
    button_mappings: Dict[str, str] = field(default_factory=dict)

    # Dead zones for axes
    dead_zones: Dict[str, float] = field(default_factory=dict)


# Example profile for Rosmaster A1 (Ackermann steering)
ROSMASTER_A1_PROFILE = RobotControlProfile(
    robot_type="rosmaster_a1",
    axis_mappings={
        "left_y": "throttle",       # Forward/back
        "left_x": "steering",       # Wheel angle
        "right_x": "camera_pan",    # Camera horizontal
        "right_y": "camera_tilt",   # Camera vertical
    },
    axis_scales={
        "throttle": 0.5,      # Max 0.5 m/s
        "steering": 30.0,     # Max 30 degrees steering
        "camera_pan": 90.0,   # 90 degree range from center
        "camera_tilt": 45.0,  # 45 degree range from center
    },
    axis_limits={
        "throttle": (-0.5, 0.5),
        "steering": (-30.0, 30.0),
        "camera_pan": (0, 180),
        "camera_tilt": (45, 135),
    },
    button_mappings={
        "a": "beep",
        "b": "emergency_stop",
        "start": "set_mode",
    },
    dead_zones={
        "throttle": 0.1,
        "steering": 0.1,
        "camera_pan": 0.05,
        "camera_tilt": 0.05,
    },
)


# ============================================================================
# VIRTUAL WORLD (Tauri Desktop App ↔ API)
# ============================================================================

class VirtualElementType(str, Enum):
    WALL = "wall"
    FRUIT = "fruit"
    NOTE = "note"
    OBSTACLE = "obstacle"


@dataclass
class VirtualWorldElement:
    """Static element in the virtual 3D world."""
    id: int = 0
    element_type: str = "wall"
    x: float = 0.0
    y: float = 0.0
    z: float = 0.0
    width: Optional[float] = None
    height: Optional[float] = None
    depth: Optional[float] = None
    rotation: float = 0.0
    data: Optional[str] = None  # JSON for element-specific data

    def to_dict(self) -> Dict:
        d = {
            "id": self.id,
            "element_type": self.element_type,
            "x": self.x,
            "y": self.y,
            "z": self.z,
            "rotation": self.rotation,
        }
        if self.width is not None:
            d["width"] = self.width
        if self.height is not None:
            d["height"] = self.height
        if self.depth is not None:
            d["depth"] = self.depth
        if self.data is not None:
            d["data"] = self.data
        return d

    @classmethod
    def from_dict(cls, d: Dict) -> "VirtualWorldElement":
        return cls(
            id=d.get("id", 0),
            element_type=d.get("element_type", "wall"),
            x=d.get("x", 0.0),
            y=d.get("y", 0.0),
            z=d.get("z", 0.0),
            width=d.get("width"),
            height=d.get("height"),
            depth=d.get("depth"),
            rotation=d.get("rotation", 0.0),
            data=d.get("data"),
        )


@dataclass
class VirtualPlayer:
    """Virtual player (cube) that can be spawned and controlled in the virtual world."""
    id: int = 0
    namespace: str = ""
    name: str = ""
    x: float = 0.0
    y: float = 0.0
    z: float = 0.5
    yaw: float = 0.0
    color: str = "#3b82f6"

    def to_dict(self) -> Dict:
        return {
            "id": self.id,
            "namespace": self.namespace,
            "name": self.name,
            "x": self.x,
            "y": self.y,
            "z": self.z,
            "yaw": self.yaw,
            "color": self.color,
        }

    @classmethod
    def from_dict(cls, d: Dict) -> "VirtualPlayer":
        return cls(
            id=d.get("id", 0),
            namespace=d.get("namespace", ""),
            name=d.get("name", ""),
            x=d.get("x", 0.0),
            y=d.get("y", 0.0),
            z=d.get("z", 0.5),
            yaw=d.get("yaw", 0.0),
            color=d.get("color", "#3b82f6"),
        )


@dataclass
class VirtualWorldState:
    """Full state of the virtual world for a lobby."""
    lobby_id: int = 0
    elements: List[VirtualWorldElement] = field(default_factory=list)
    players: List[VirtualPlayer] = field(default_factory=list)

    def to_dict(self) -> Dict:
        return {
            "lobby_id": self.lobby_id,
            "elements": [e.to_dict() for e in self.elements],
            "players": [p.to_dict() for p in self.players],
        }

    @classmethod
    def from_dict(cls, d: Dict) -> "VirtualWorldState":
        return cls(
            lobby_id=d.get("lobby_id", 0),
            elements=[VirtualWorldElement.from_dict(e) for e in d.get("elements", [])],
            players=[VirtualPlayer.from_dict(p) for p in d.get("players", [])],
        )
